File: packages/core/src/KeyRegistry.test.ts
import { KeyRegistry } from './KeyRegistry.js';
import type { StorageKeyDef } from './types.js';
import { Scope } from './types.js';

const keyDef = (key: string, scope: Scope): StorageKeyDef => ({ key, scope });

describe('Scope enum', () => {
  it('has value "user" for User', () => {
    expect(Scope.User).toBe('user');
  });

  it('has value "device" for Device', () => {
    expect(Scope.Device).toBe('device');
  });

  it('has value "global" for Global', () => {
    expect(Scope.Global).toBe('global');
  });

  it('has exactly 3 members', () => {
    expect(Object.values(Scope)).toHaveLength(3);
  });
});

describe('KeyRegistry', () => {
  let registry: KeyRegistry;

  beforeEach(() => {
    registry = new KeyRegistry();
  });

  describe('userId management', () => {
    it('returns null initially', () => {
      expect(registry.getUserId()).toBeNull();
    });

    it('stores the userId after setUserId()', () => {
      registry.setUserId('alice');
      expect(registry.getUserId()).toBe('alice');
    });

    it('allows changing userId with subsequent setUserId() calls', () => {
      registry.setUserId('alice');
      registry.setUserId('bob');
      expect(registry.getUserId()).toBe('bob');
    });

    it('resets userId to null after clearUserId()', () => {
      registry.setUserId('alice');
      registry.clearUserId();
      expect(registry.getUserId()).toBeNull();
    });

    it('does not throw when clearUserId() is called without a userId set', () => {
      expect(() => registry.clearUserId()).not.toThrow();
      expect(registry.getUserId()).toBeNull();
    });
  });

  describe('resolve()', () => {
    describe('Scope.User', () => {
      it('resolves to u/{userId}/{key}', () => {
        registry.setUserId('alice');
        expect(registry.resolve(keyDef('theme', Scope.User))).toBe('u/alice/theme');
      });

      it('throws when userId is not set', () => {
        expect(() => registry.resolve(keyDef('theme', Scope.User))).toThrow(
          'Keyward: userId not set.',
        );
      });

      it('throws after userId is cleared', () => {
        registry.setUserId('alice');
        registry.clearUserId();
        expect(() => registry.resolve(keyDef('theme', Scope.User))).toThrow(
          'Keyward: userId not set.',
        );
      });

      it('uses the current userId after change', () => {
        registry.setUserId('alice');
        expect(registry.resolve(keyDef('theme', Scope.User))).toBe('u/alice/theme');
        registry.setUserId('bob');
        expect(registry.resolve(keyDef('theme', Scope.User))).toBe('u/bob/theme');
      });
    });

    describe('Scope.Device', () => {
      it('resolves to d/{key}', () => {
        expect(registry.resolve(keyDef('volume', Scope.Device))).toBe('d/volume');
      });

      it('does not include userId even when one is set', () => {
        registry.setUserId('alice');
        expect(registry.resolve(keyDef('volume', Scope.Device))).toBe('d/volume');
      });
    });

    describe('Scope.Global', () => {
      it('resolves to the bare key', () => {
        expect(registry.resolve(keyDef('app-version', Scope.Global))).toBe('app-version');
      });

      it('does not include userId even when one is set', () => {
        registry.setUserId('alice');
        expect(registry.resolve(keyDef('app-version', Scope.Global))).toBe('app-version');
      });
    });

    describe('edge cases', () => {
      it('handles empty string key for Scope.Global', () => {
        expect(registry.resolve(keyDef('', Scope.Global))).toBe('');
      });

      it('handles empty string key for Scope.Device', () => {
        expect(registry.resolve(keyDef('', Scope.Device))).toBe('d/');
      });

      it('handles empty string key for Scope.User', () => {
        registry.setUserId('alice');
        expect(registry.resolve(keyDef('', Scope.User))).toBe('u/alice/');
      });

      it('handles key with slashes', () => {
        expect(registry.resolve(keyDef('a/b/c', Scope.Device))).toBe('d/a/b/c');
      });

      it('handles key with special characters', () => {
        expect(registry.resolve(keyDef('settings.theme@v2', Scope.Global))).toBe(
          'settings.theme@v2',
        );
      });

      it('handles userId with special characters', () => {
        registry.setUserId('user@example.com');
        expect(registry.resolve(keyDef('prefs', Scope.User))).toBe('u/user@example.com/prefs');
      });

      it('handles key with spaces', () => {
        expect(registry.resolve(keyDef('my key', Scope.Device))).toBe('d/my key');
      });

      it('handles key with unicode characters', () => {
        registry.setUserId('alice');
        expect(registry.resolve(keyDef('tema-\u00e7', Scope.User))).toBe('u/alice/tema-\u00e7');
        expect(registry.resolve(keyDef('tema-\u00e7', Scope.Device))).toBe('d/tema-\u00e7');
        expect(registry.resolve(keyDef('tema-\u00e7', Scope.Global))).toBe('tema-\u00e7');
      });
    });
  });

  describe('getUserPrefix()', () => {
    it('returns u/{userId}/', () => {
      expect(registry.getUserPrefix('alice')).toBe('u/alice/');
    });

    it('works with special characters in userId', () => {
      expect(registry.getUserPrefix('user@example.com')).toBe('u/user@example.com/');
    });

    it('is independent of currentUserId state', () => {
      expect(registry.getUserPrefix('bob')).toBe('u/bob/');
    });

    it('prefix matches resolve output for User-scoped keys', () => {
      registry.setUserId('alice');
      const resolved = registry.resolve(keyDef('theme', Scope.User));
      expect(resolved.startsWith(registry.getUserPrefix('alice'))).toBe(true);
    });
  });

  describe('instance isolation', () => {
    it('separate instances do not share state', () => {
      const other = new KeyRegistry();
      registry.setUserId('alice');
      other.setUserId('bob');
      expect(registry.resolve(keyDef('k', Scope.User))).toBe('u/alice/k');
      expect(other.resolve(keyDef('k', Scope.User))).toBe('u/bob/k');
    });
  });
});

File: packages/platform-android/src/test/java/org/keyward/KeyRegistryTest.java
package org.keyward;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

final class KeyRegistryTest {

    private KeyRegistry registry;

    @BeforeEach
    void setUp() {
        registry = new KeyRegistry();
    }

    // Scope enum

    @Test
    void scopeHasThreeCases() {
        assertEquals(3, Scope.values().length);
    }

    // userId management

    @Test
    void returnsNullInitially() {
        assertNull(registry.getUserId());
    }

    @Test
    void storesUserIdAfterSetUserId() {
        registry.setUserId("alice");
        assertEquals("alice", registry.getUserId());
    }

    @Test
    void allowsChangingUserIdWithSubsequentCalls() {
        registry.setUserId("alice");
        registry.setUserId("bob");
        assertEquals("bob", registry.getUserId());
    }

    @Test
    void resetsUserIdToNullAfterClearUserId() {
        registry.setUserId("alice");
        registry.clearUserId();
        assertNull(registry.getUserId());
    }

    @Test
    void doesNotThrowWhenClearUserIdCalledWithoutUserId() {
        registry.clearUserId();
        assertNull(registry.getUserId());
    }

    // resolve: Scope.USER

    @Test
    void userScopeResolvesToPrefixedKey() throws KeywardException {
        registry.setUserId("alice");
        assertEquals("u/alice/theme", registry.resolve(new KeyDef("theme", Scope.USER)));
    }

    @Test
    void userScopeThrowsWhenUserIdNotSet() {
        KeywardException ex = assertThrows(KeywardException.class,
                () -> registry.resolve(new KeyDef("theme", Scope.USER)));
        assertEquals("Keyward: userId not set.", ex.getMessage());
    }

    @Test
    void userScopeThrowsAfterUserIdCleared() {
        registry.setUserId("alice");
        registry.clearUserId();
        assertThrows(KeywardException.class,
                () -> registry.resolve(new KeyDef("theme", Scope.USER)));
    }

    @Test
    void userScopeUsesCurrentUserIdAfterChange() throws KeywardException {
        registry.setUserId("alice");
        assertEquals("u/alice/theme", registry.resolve(new KeyDef("theme", Scope.USER)));
        registry.setUserId("bob");
        assertEquals("u/bob/theme", registry.resolve(new KeyDef("theme", Scope.USER)));
    }

    // resolve: Scope.DEVICE

    @Test
    void deviceScopeResolvesToPrefixedKey() throws KeywardException {
        assertEquals("d/volume", registry.resolve(new KeyDef("volume", Scope.DEVICE)));
    }

    @Test
    void deviceScopeDoesNotIncludeUserId() throws KeywardException {
        registry.setUserId("alice");
        assertEquals("d/volume", registry.resolve(new KeyDef("volume", Scope.DEVICE)));
    }

    // resolve: Scope.GLOBAL

    @Test
    void globalScopeResolvesToBareKey() throws KeywardException {
        assertEquals("app-version", registry.resolve(new KeyDef("app-version", Scope.GLOBAL)));
    }

    @Test
    void globalScopeDoesNotIncludeUserId() throws KeywardException {
        registry.setUserId("alice");
        assertEquals("app-version", registry.resolve(new KeyDef("app-version", Scope.GLOBAL)));
    }

    // resolve: edge cases

    @Test
    void emptyStringKeyGlobal() throws KeywardException {
        assertEquals("", registry.resolve(new KeyDef("", Scope.GLOBAL)));
    }

    @Test
    void emptyStringKeyDevice() throws KeywardException {
        assertEquals("d/", registry.resolve(new KeyDef("", Scope.DEVICE)));
    }

    @Test
    void emptyStringKeyUser() throws KeywardException {
        registry.setUserId("alice");
        assertEquals("u/alice/", registry.resolve(new KeyDef("", Scope.USER)));
    }

    @Test
    void keyWithSlashes() throws KeywardException {
        assertEquals("d/a/b/c", registry.resolve(new KeyDef("a/b/c", Scope.DEVICE)));
    }

    @Test
    void keyWithSpecialCharacters() throws KeywardException {
        assertEquals("settings.theme@v2", registry.resolve(new KeyDef("settings.theme@v2", Scope.GLOBAL)));
    }

    @Test
    void userIdWithSpecialCharacters() throws KeywardException {
        registry.setUserId("user@example.com");
        assertEquals("u/user@example.com/prefs", registry.resolve(new KeyDef("prefs", Scope.USER)));
    }

    @Test
    void keyWithSpaces() throws KeywardException {
        assertEquals("d/my key", registry.resolve(new KeyDef("my key", Scope.DEVICE)));
    }

    @Test
    void keyWithUnicodeCharacters() throws KeywardException {
        registry.setUserId("alice");
        assertEquals("u/alice/tema-\u00e7", registry.resolve(new KeyDef("tema-\u00e7", Scope.USER)));
        assertEquals("d/tema-\u00e7", registry.resolve(new KeyDef("tema-\u00e7", Scope.DEVICE)));
        assertEquals("tema-\u00e7", registry.resolve(new KeyDef("tema-\u00e7", Scope.GLOBAL)));
    }

    // getUserPrefix

    @Test
    void getUserPrefixFormat() {
        assertEquals("u/alice/", KeyRegistry.getUserPrefix("alice"));
    }

    @Test
    void getUserPrefixWithSpecialCharacters() {
        assertEquals("u/user@example.com/", KeyRegistry.getUserPrefix("user@example.com"));
    }

    @Test
    void getUserPrefixIsIndependentOfState() {
        assertEquals("u/bob/", KeyRegistry.getUserPrefix("bob"));
    }

    @Test
    void getUserPrefixMatchesResolveOutput() throws KeywardException {
        registry.setUserId("alice");
        String resolved = registry.resolve(new KeyDef("theme", Scope.USER));
        assertTrue(resolved.startsWith(KeyRegistry.getUserPrefix("alice")));
    }

    // instance isolation

    @Test
    void separateInstancesDoNotShareState() throws KeywardException {
        KeyRegistry other = new KeyRegistry();
        registry.setUserId("alice");
        other.setUserId("bob");
        assertEquals("u/alice/k", registry.resolve(new KeyDef("k", Scope.USER)));
        assertEquals("u/bob/k", other.resolve(new KeyDef("k", Scope.USER)));
    }
}
